File: binding/src/libqrunnable.cpp
#include <QRunnable>
#include <qrunnable.h>
#include "libqrunnable.h"
#include "libqrunnable.hxx"

QRunnable* QRunnable_new() {
	 return new VirtualQRunnable();
}

void QRunnable_Run(QRunnable* self) {
	self->run();
}

bool QRunnable_AutoDelete(const QRunnable* self) {
	return self->autoDelete();
}

void QRunnable_SetAutoDelete(QRunnable* self, bool autoDelete) {
	self->setAutoDelete(autoDelete);
}

// Base class handler implementation
void QRunnable_QBaseRun(QRunnable* self) {
	auto* vqrunnable = dynamic_cast<VirtualQRunnable*>(self);
	if (vqrunnable && vqrunnable->isVirtualQRunnable) {
vqrunnable->setQRunnable_Run_IsBase(true);
	vqrunnable->run();
}
}

// Auxiliary method to allow providing re-implementation
void QRunnable_OnRun(QRunnable* self, intptr_t slot) {
	auto* vqrunnable = dynamic_cast<VirtualQRunnable*>(self);
	if (vqrunnable && vqrunnable->isVirtualQRunnable) {
vqrunnable->setQRunnable_Run_Callback(reinterpret_cast<VirtualQRunnable::QRunnable_Run_Callback>(slot));
}
}

void QRunnable_Delete(QRunnable* self) {
    delete self;
}

void QGenericRunnable_Run(QGenericRunnable* self) {
	self->run();
}

// Base class handler implementation
void QGenericRunnable_QBaseRun(QGenericRunnable* self) {
	auto* vqgenericrunnable = dynamic_cast<VirtualQGenericRunnable*>(self);
	if (vqgenericrunnable && vqgenericrunnable->isVirtualQGenericRunnable) {
vqgenericrunnable->setQGenericRunnable_Run_IsBase(true);
	vqgenericrunnable->run();
}
}

// Auxiliary method to allow providing re-implementation
void QGenericRunnable_OnRun(QGenericRunnable* self, intptr_t slot) {
	auto* vqgenericrunnable = dynamic_cast<VirtualQGenericRunnable*>(self);
	if (vqgenericrunnable && vqgenericrunnable->isVirtualQGenericRunnable) {
vqgenericrunnable->setQGenericRunnable_Run_Callback(reinterpret_cast<VirtualQGenericRunnable::QGenericRunnable_Run_Callback>(slot));
}
}

void QGenericRunnable_Delete(QGenericRunnable* self) {
    delete self;
}

File: binding/src/libqrunnable.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qrunnable.html
pub const qrunnable = struct {

    /// New constructs a new QRunnable object.
    pub fn New() QtC.QRunnable {
        return qtc.QRunnable_new();
    }


    pub fn Run(self: ?*anyopaque, ) void {
        qtc.QRunnable_Run(@ptrCast(self));
    }

    /// Allows for overriding the related default method
    pub fn OnRun(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QRunnable_OnRun(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseRun(self: ?*anyopaque, ) void {
        qtc.QRunnable_QBaseRun(@ptrCast(self));
    }

    pub fn AutoDelete(self: ?*anyopaque, ) bool {
        return qtc.QRunnable_AutoDelete(@ptrCast(self));
    }

    pub fn SetAutoDelete(self: ?*anyopaque, autoDelete: bool) void {
        qtc.QRunnable_SetAutoDelete(@ptrCast(self), autoDelete);
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QRunnable_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/qgenericrunnable.html
pub const qgenericrunnable = struct {

    pub fn Run(self: ?*anyopaque, ) void {
        qtc.QGenericRunnable_Run(@ptrCast(self));
    }

    /// Allows for overriding the related default method
    pub fn OnRun(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QGenericRunnable_OnRun(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseRun(self: ?*anyopaque, ) void {
        qtc.QGenericRunnable_QBaseRun(@ptrCast(self));
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QGenericRunnable_Delete(@ptrCast(self));
    }
};
